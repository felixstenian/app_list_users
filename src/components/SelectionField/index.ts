export { default } from './SelectionField'
export type { SelectionFieldProps } from './SelectionField'
