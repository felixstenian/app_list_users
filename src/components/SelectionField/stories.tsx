import { Meta, Story } from '@storybook/react'
import SelectionField, { SelectionFieldProps } from '.'

export default {
  title: 'SelectionField',
  component: SelectionField,
  args: {
    options: {
      type: Array<string>
    }
  }
} as Meta

export const Default: Story<SelectionFieldProps> = (args) => (
  <SelectionField {...args} />
)

Default.args = {
  options: ['Option 1', 'Option 2']
}
