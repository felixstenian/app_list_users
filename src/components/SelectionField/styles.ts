import styled, { css } from 'styled-components'
import { layout } from 'styled-system'

export const Label = styled.label`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.small};
    color: ${theme.colors.black};
    padding: ${theme.spacings.xxsmall};
  `}
`

export const InputWraper = styled.div`
  display: flex;
  align-items: center;
  border-radius: 4rem;
  border: 0.2rem solid;
  width: 100%;

  ${({ theme }) => css`
    background: ${theme.colors.lightGray};
    padding: ${theme.spacings.xxsmall};
    border-color: ${theme.colors.lightGray};

    &:focus-within {
      box-shadow: 0 0 0.5rem ${theme.colors.gray};
    }
  `}
`

export const Input = styled.select`
  ${({ theme }) => css`
    color: ${theme.colors.black};
    font-family: ${theme.font.family};
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacings.xxsmall} 0;
    background: transparent;
    border: 0;
    outline: none;
    cursor: pointer;
  `}
  ${layout};
`
export const Option = styled.option`
  ${({ theme }) => css`
    background: ${theme.colors.lightGray};
    border: 0;
    cursor: pointer;
  `}
`
