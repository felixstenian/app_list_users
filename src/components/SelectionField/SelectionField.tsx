import { InputHTMLAttributes, useState } from 'react'

import Row from 'components/Row/Row'
import Text from 'components/Text/Text'
import { LayoutProps } from 'styled-system'
import * as Style from './styles'

export type SelectionFieldProps = {
  label?: string
  placeholder?: string
  initialValue?: string
  options: string[]
} & InputHTMLAttributes<HTMLSelectElement> &
  LayoutProps

const SelectionField = ({
  label,
  placeholder,
  initialValue,
  name,
  options,
  ...props
}: SelectionFieldProps) => {
  const [selectedOption, setSelectedOption] = useState(
    initialValue || options[0]
  )

  return (
    <Row alignItems="center">
      {!!label && <Style.Label htmlFor={name}>{label}</Style.Label>}
      <Style.InputWraper>
        <Style.Input
          onChange={(e) => setSelectedOption(e.target.value)}
          value={selectedOption}
          name={name}
          placeholder={placeholder}
          {...(label ? { id: name } : {})}
          {...props}
        >
          {options.map((option) => (
            <Style.Option key={option} value={option}>
              <Row>
                <Text>{option}</Text>
              </Row>
            </Style.Option>
          ))}
        </Style.Input>
      </Style.InputWraper>
    </Row>
  )
}

export default SelectionField
