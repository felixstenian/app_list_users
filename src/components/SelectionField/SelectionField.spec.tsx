import { screen } from '@testing-library/react'

import { renderWithTheme } from 'utils'
import SelectionField from '.'

const options = ['Option 1', 'Option 2']

describe('<SelectionField />', () => {
  it('1. should render the reading', () => {
    const { container } = renderWithTheme(
      <SelectionField label="SelectionField" options={options} />
    )

    expect(screen.getByText(/SelectionField/i)).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })
  it('2. should render the label', () => {
    const label = 'Label'
    renderWithTheme(<SelectionField label={label} options={options} />)
    expect(screen.getByText(label)).toBeInTheDocument()
  })
  it('3. should render the placeholder', () => {
    const placeholder = 'Placeholder'
    renderWithTheme(
      <SelectionField
        role="select"
        placeholder={placeholder}
        options={options}
      />
    )

    expect(screen.getByRole('select')).toHaveAttribute(
      'placeholder',
      placeholder
    )
  })
  it('4. should render the options', () => {
    renderWithTheme(<SelectionField options={options} />)
    expect(screen.getByText(options[0])).toBeInTheDocument()
    expect(screen.getByText(options[1])).toBeInTheDocument()
  })
})
