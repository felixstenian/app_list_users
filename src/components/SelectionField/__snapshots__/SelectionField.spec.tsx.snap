// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectionField /> 1. should render the reading 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c6 {
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.4rem;
}

.c1 {
  font-size: 1.4rem;
  color: #030517;
  padding: 0.8rem;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 4rem;
  border: 0.2rem solid;
  width: 100%;
  background: #EAEAEA;
  padding: 0.8rem;
  border-color: #EAEAEA;
}

.c2:focus-within {
  box-shadow: 0 0 0.5rem #8F8F8F;
}

.c3 {
  color: #030517;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.6rem;
  padding: 0.8rem 0;
  background: transparent;
  border: 0;
  outline: none;
  cursor: pointer;
}

.c4 {
  background: #EAEAEA;
  border: 0;
  cursor: pointer;
}

<div
  class="c0"
>
  <label
    class="c1"
  >
    SelectionField
  </label>
  <div
    class="c2"
  >
    <select
      class="c3"
    >
      <option
        class="c4"
        value="Option 1"
      >
        <div
          class="c5"
        >
          <body
            class="c6"
          >
            Option 1
          </body>
        </div>
      </option>
      <option
        class="c4"
        value="Option 2"
      >
        <div
          class="c5"
        >
          <body
            class="c6"
          >
            Option 2
          </body>
        </div>
      </option>
    </select>
  </div>
</div>
`;
